### PR TITLE
Skip ready queue when no resources

### PR DESCRIPTION
**Pull request by istoica**
## What do these changes do?

This PR improves the performance of selecting a task from ready queue to schedule when the resource availability change or a new worker becomes available. It is doing so by maintaining (1) the minimum resource quantities required by any task in the ready queue, and (2) and the number of such tasks. If (1) is larger than the available resources at the node, there is no need to walk through the ready queue, as we know that no tasks can be scheduled. 

## Related issue number

 Resolves the issue https://github.com/ray-project/ray/issues/3194.

### CODE REVIEW

**Review by pcmoritz — APPROVED** [review]
Can be merged when tests pass

**Review by pcmoritz — COMMENTED** [review]
refactoring this now

File: src/ray/raylet/scheduling_queue.cc [hunk 1]
```diff
@@ -73,111 +76,107 @@ namespace ray {
...
-  return local_available_resources.Contains(ready_tasks_metadata_.GetMinTaskResources());
+void ReadyQueue::UpdateMinOnRemove(const ResourceSet &resources) {
+  if (min_task_resources_ == resources) {
+    min_task_count_--;
+  }
 };
 
+void ReadyQueue::RecomputeMinResources() {
+  for (const auto &task : task_list_) {
+    UpdateMinOnAdd(task.GetTaskSpecification().GetRequiredResources());
```
Comment (atumanov) [inline src/ray/raylet/scheduling_queue.cc]: we can speed this up to be O(logn) by maintaining a min heap

File: src/ray/raylet/scheduling_queue.h [hunk 2]
```diff
@@ -37,6 +37,118 @@ enum class TaskState {
...
+  /// \return Whether the task_id exists in this queue.
+  bool HasTask(const TaskID &task_id) const;
+
+  /// \brief Remove the task list of the queue.
+  /// \return A list of tasks contained in this queue.
+  const std::list<Task> &GetTasks() const;
+
+  /// \brief Recompute the resources of the tasks with minimum requirements
+  /// in the ready queue.
+  void RecomputeMinResources();
```
Comment (atumanov) [inline src/ray/raylet/scheduling_queue.h]: we don't need `RecomputeMinResources` in the `TaskQueue` do we.

File: src/ray/raylet/scheduling_queue.h [hunk 3]
```diff
@@ -37,6 +37,118 @@ enum class TaskState {
...
+                  std::vector<Task> *removed_tasks = nullptr);
+
+  /// \brief Check whether any of the tasks in the ready queue can be scheduled.
+  ///
+  /// \param local_available_resources: resources available on local node.
+  /// \return True if there is at least a task that can be scheduled, i.e.
+  /// whose requirements can be satisfied by the local node; false otherwise.
+  const bool CanScheduleMinTask(const ResourceIdSet &local_available_resources) const;
+
+  const ResourceSet &GetCurrentResourceLoad() const { return current_resource_load_; }
```
Comment (atumanov) [inline src/ray/raylet/scheduling_queue.h]: `current_resource_load_` should ideally be a field of the `TaskQueue` parent, not just the child class. That way, we can easily add more queues to the definition of "load".

File: src/ray/raylet/scheduling_queue.h [hunk 4]
```diff
@@ -37,6 +37,118 @@ enum class TaskState {
...
+  /// ready queue.
+  ///
+  /// \return Number of tasks with minimum resource requirements.
+  const int &GetMinTaskCount() const { return min_task_count_;};
+
+  /// \brief Update the min  resources required by a task in the ready queue,
+  ///  when a new task is added to the queue.
+  ///
+  /// \param resources: Resource requirements of the new task being added.
+  void UpdateMinOnAdd(const ResourceSet &resources);
```
Comment (atumanov) [inline src/ray/raylet/scheduling_queue.h]: `UpdateMinOnAdd` and `UpdateMinOnRemove` seem like private methods to me.

File: src/ray/raylet/scheduling_queue.h [hunk 5]
```diff
@@ -37,6 +37,118 @@ enum class TaskState {
...
+
+  /// \brief Update the min resources required by a task in the ready queue,
+  ///  when a task is removed from the queue.
+  ///
+  /// \param resources: Resource requirements of the new task being removed.
+  void UpdateMinOnRemove(const ResourceSet &resources);
+
+  /// \brief Recompute the resources of the tasks with minimum requirements
+  /// in the ready queue.
+  void RecomputeMinResources();
```
Comment (atumanov) [inline src/ray/raylet/scheduling_queue.h]: `RecomputeMinResources` probably also private

File: src/ray/raylet/scheduling_queue.h [hunk 6]
```diff
@@ -37,6 +37,118 @@ enum class TaskState {
...
+
+private:
+  /// Aggregate resources of all the tasks in this queue.
+  ResourceSet current_resource_load_;
+  /// Minimum resources requested by any task in tyhe ready queue.
+  ResourceSet min_task_resources_;
+  /// Count of tasks which the minimum resource requirements in the ready queue.
+  int min_task_count_ = 0;
```
Comment (atumanov) [inline src/ray/raylet/scheduling_queue.h]: We should initialize these in `ReadyQueue::ReadyQueue()` ctor

File: src/ray/raylet/scheduling_queue.cc [hunk 7]
```diff
@@ -73,111 +76,107 @@ namespace ray {
...
+TaskQueue::~TaskQueue() {
   task_map_.clear();
   task_list_.clear();
 }
 
-bool SchedulingQueue::TaskQueue::AppendTask(const TaskID &task_id,
-                                            const Task &task,
-                                            bool is_ready_queue) {
+bool TaskQueue::AppendTask(const TaskID &task_id,
+                                            const Task &task) {
```
[the remark this answers is not in the document]
Comment (pcmoritz) [reply src/ray/raylet/scheduling_queue.cc]: Yeah, I haven't done the linting yet, will do that as soon as travis finishes :)

File: src/ray/raylet/scheduling_queue.h [hunk 8]
```diff
@@ -37,44 +37,67 @@ enum class TaskState {
...
+  void RecomputeMinResources();
+
+protected:
+ /// A list of tasks.
+ std::list<Task> task_list_;
+ /// A hash to speed up looking up a task.
+ std::unordered_map<TaskID, std::list<Task>::iterator> task_map_;
+};
+
+class ReadyQueue : public TaskQueue {
```
[the remark this answers is not in the document]
Comment (pcmoritz) [reply src/ray/raylet/scheduling_queue.h]: ok sounds good! TaskQueueWithStats would be a good name I think.

File: src/ray/raylet/scheduling_queue.cc (hunk 1 above)
Comment (pcmoritz) [reply src/ray/raylet/scheduling_queue.cc]: Yeah, good idea! I'll leave it as a separate PR for now.

File: src/ray/raylet/scheduling_queue.h (hunk 2 above)
Comment (pcmoritz) [reply src/ray/raylet/scheduling_queue.h]: Oh yeah, that's a leftover, will remove. Thanks for catching it.

File: src/ray/raylet/scheduling_queue.h (hunk 3 above)
Comment (pcmoritz) [reply src/ray/raylet/scheduling_queue.h]: Ok yeah. That might actually be a good candidate for a flag as it might be useful for any queue, I can look into it.

File: src/ray/raylet/scheduling_queue.h (hunk 4 above)
Comment (pcmoritz) [reply src/ray/raylet/scheduling_queue.h]: Yeah, I did that first, but they are used in the tests. There is a way to make the test a friend, but that might be overkill (if it's easy to do I can do it).

File: src/ray/raylet/scheduling_queue.h (hunk 5 above)
Comment (pcmoritz) [reply src/ray/raylet/scheduling_queue.h]: Same as above

File: src/ray/raylet/scheduling_queue.h (hunk 6 above)
Comment (pcmoritz) [reply src/ray/raylet/scheduling_queue.h]: Hm, not clear that that buys anything here. The ResourceSet constructor will already be automatically called and the min_task_count_ is initialized.

File: src/ray/raylet/scheduling_queue.h [hunk 9]
```diff
@@ -37,6 +37,118 @@ enum class TaskState {
...
+  /// \return Whether the removal succeeds.
+  bool RemoveTask(const TaskID &task_id,
+                  std::vector<Task> *removed_tasks = nullptr);
+
+  /// \brief Check whether any of the tasks in the ready queue can be scheduled.
+  ///
+  /// \param local_available_resources: resources available on local node.
+  /// \return True if there is at least a task that can be scheduled, i.e.
+  /// whose requirements can be satisfied by the local node; false otherwise.
+  const bool CanScheduleMinTask(const ResourceIdSet &local_available_resources) const;
```
Comment (stephanie-wang) [inline src/ray/raylet/scheduling_queue.h]: This method seems unnecessary since you can always do this from the caller by calling `GetMinTaskResources()`

File: src/ray/raylet/scheduling_queue.h (hunk 4 above)
Comment (stephanie-wang) [reply src/ray/raylet/scheduling_queue.h]: I think it is easy and preferable to make the test class a friend.

File: src/ray/raylet/scheduling_queue.cc [hunk 10]
```diff
@@ -6,7 +6,8 @@ namespace {
 
 // Helper function to remove tasks in the given set of task_ids from a
 // queue, and append them to the given vector removed_tasks.
-void RemoveTasksFromQueue(ray::raylet::SchedulingQueue::TaskQueue &queue,
+template<typename TaskQueue>
+void RemoveTasksFromQueue(TaskQueue &queue,
```
Comment (stephanie-wang) [inline src/ray/raylet/scheduling_queue.cc]: Dumb question, but why is this necessary?

File: src/ray/raylet/scheduling_queue.h [hunk 11]
```diff
@@ -37,6 +37,118 @@ enum class TaskState {
...
+};
+
+class ReadyQueue : public TaskQueue {
+public:
+  /// \brief Append a task to the ready queue.
+  ///
+  /// \param task_id The task ID for the task to append.
+  /// \param task The task to append to the queue.
+  /// \return Whether the append operation succeeds.
+  bool AppendTask(const TaskID &task_id, const Task &task);
```
Comment (stephanie-wang) [inline src/ray/raylet/scheduling_queue.h]: Add `override` to these methods?

**Review by stephanie-wang — COMMENTED** [review]
I wonder if a simpler design is to keep a mapping between resource maps -> count of tasks that require those resources? Usually the number of unique resource maps will probably be pretty low, so iterating through it to find the lowest will be constant time. When there are many many unique resource maps in the queue, it will be about the same complexity as the current PR since you will often have to recompute the minimum resources when tasks are removed.

File: src/ray/raylet/scheduling_queue.cc (hunk 10 above)
Comment (atumanov) [reply src/ray/raylet/scheduling_queue.cc]: templating or the method itself? The method itself, IMHO, should be made a method of TaskQueue: `TaskQueue::RemoveTasks` to supplement `TaskQueue::RemoveTask`. Templating here is necessary, because the child class ReadyQueue is not a polymorphic type. So calling any method that takes `TaskQueue` as an argument will result in calling the parent class's implementation. 
Finally, why not use virtual? Great question! There's concern about performance implications of using virtual functions. The current PR design will make it straightforward to go from vanilla subclassing to polymorphic subclassing.

File: src/ray/raylet/scheduling_queue.h (hunk 11 above)
Comment (atumanov) [reply src/ray/raylet/scheduling_queue.h]: I think you only need it for virtual functions:
https://en.cppreference.com/w/cpp/language/override

File: src/ray/raylet/scheduling_queue.h (hunk 9 above)
Comment (atumanov) [reply src/ray/raylet/scheduling_queue.h]: in the current implementation, true. But as this needs to handle the multi-dimensional (`k-d`) resource space, I actually think we will need a function like this. The reason for this is that a single `ResourceSet min_task_resources_` will not be enough. There's a bit more bookkeeping that should happen, because we're effectively doing an "orthogonal range query" here.

File: src/ray/raylet/scheduling_queue.h (hunk 4 above)
Comment (atumanov) [reply src/ray/raylet/scheduling_queue.h]: I agree. We should keep interfaces as lean and orthogonal as possible.

**Comment by pcmoritz** [comment]
@stephanie-wang @ericl suggested a related idea, to have one subqueue per resource shape so we don't need to keep track of any minima etc. and DispatchTask can be O(1). It can actually be implemented as an additional index to our current queue and I'm trying that out at the moment.

**Comment by stephanie-wang** [comment]
Sounds good. We might want to just add the task counts instead of another index to keep the code simpler, but if it works out then let's try it.

**Comment by atumanov** [comment]
I really don't think it's a good idea to assume that the number of resource shapes is going to be a small constant. This may be true for most (cpu, gpu) cases (even then, 64x8 or more in the worst case). But it may easily blow up when people start using custom resources.
Splitting up a queue into multiple queues will also lose the FIFO order.

Here's what I proposed for this PR/release to keep things simple:
Check whether \vec{avail_res} intersects the value range in each dimension. It’s a necessary but not sufficient condition for the existence of a scheduleable task that fits within \vec{avail_res}. If there exists a resource dimension i , such that avail_res_i < min_res_i , we know that no task can be scheduled. This is going to take care of 90% of cases, because most of the cases we care about are CPU cases and exiting early if there are no zero CPU tasks queued...

**Comment by atumanov** [comment]
I'm sorry, but I'm against splitting up the queue just to keep track of resource shapes, this is just bad engineering and is a stop-gap measure that will blow up with dynamic custom resources (which can be with capacity of infinity, by the way).

**Comment by atumanov** [comment]
going forward, IF we need something more sophisticated, we may think about this problem as a k-d range search (aka orthogonal range query), which has known solutions. We could also relax the k-d search problem and consider a slab-allocator approach idea, where we keep track of tasks in powers of two buckets. The intuition there will be that the number of tasks with "odd" sizes is going to exponentially decay, as we exponentially increase the bucket size, keeping the number of tasks in each bucket contained. Most of the tasks will be in the smallest buckets. 

For this PR: my proposition is to just do the simple test to see if available resources are less than the minimum tracked resource along any of the dimensions. This can easily be done before the code freeze and has good use case coverage.

**Comment by ericl** [comment]
@atumanov that seems like a nice optimization to have in any case. From a release perspective, I think we can go with that or Phillip's PR or both, depending on timing (however I don't see an implemention of the former yet). I also think you can do approximate fifo with multiple queues straightforwardly.
